### PR TITLE
[Serial] Modified Serial binding for ON/OFF commands

### PR DESCRIPTION
**Pull request by marekhalmo**
Modified Serial binding for ON/OFF commands to be used in conjunction with Switch item (faster). 
This can also be used with a Switch configured as a button (using mappings)

### CODE REVIEW

File: bundles/binding/org.openhab.binding.serial/README.md [hunk 1]
```diff
@@ -49,12 +49,13 @@ This binding does not have a configuration.
 
 ## Item Configuration
 
-The format has three variations:
+The format has following variations:
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.serial/README.md]: "The format has the following variations:"

File: bundles/binding/org.openhab.binding.serial/README.md [hunk 2]
```diff
@@ -63,6 +64,7 @@ where:
 * `<baudrate>` is the baud rate of the port. Backward compatibility is given; if no baud rate is specified  the serial binding defaults to 9600 baud.
 * `REGEX(<regular expression>)` allows parsing for special strings or numbers in the serial stream. This is based on the [RegEx Service](https://github.com/openhab/openhab1-addons/wiki/Transformations#regex-transformation-service). This is optional. 
 * `BASE64` enables the Base64 mode. With this mode all data received on the serial port is saved in Base64 format. In this mode also all data that is sent to the serial port has to be Base64 encoded. (This was implemented because some serial devices are using bytes that are not supported by the REST interface). 
+* `ON(<On string>),OFF(<Off string>)` if used with conjunction to Switch this mapping will send specific commands to serial port and also match a serial command to specific ON/OFF state. This way you don't have to use a rule to send a command to serial
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.serial/README.md]: "if used in conjunction with a Switch, "

File: bundles/binding/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/SerialBinding.java [hunk 3]
```diff
@@ -145,25 +152,30 @@ public void validateItemType(Item item, String bindingConfig) throws BindingConf
...
+        if (split.length > 0) {
+            for (int i = 1; i < split.length; i++) {
+                String part = split[i];
+                String substring = part.substring(0, part.length());
+
+                if (substring.startsWith("REGEX(")) {
+                    pattern = substring.substring(6, substring.length() - 1);
+                } else if (substring.equals("BASE64")) {
+                    base64 = true;
+                } else if (substring.startsWith("ON(")) {
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/SerialBinding.java]: These substring matches are error-prone. You should consider beefing up the parsing.

However, I recognize you've just expanded on what already existed here, so this is not required.

**Comment by marekhalmo** [comment]
Fixed + added Rollershutter support.

**Comment by marekhalmo** [comment]
Could you help me ho i can debug the REGEX transformation service in development?
I get this error when launching:

> !MESSAGE Could not bind a reference of component org.openhab.binding.serial. The reference is: Reference[name = TransformationService, interface = org.openhab.core.transform.TransformationService, policy = dynamic, cardinality = 0..1, target = (openhab.transform=REGEX), bind = setTransformationService, unbind = unsetTransformationService]

**Comment by marekhalmo** [comment]
Just a note to the regex transformation debug. I added it to plugins in Debug configuration. I cant see a Transformations tab in Paper UI. I tried to test the yahoo binding to check if the ESH project works partially.. and it works.. And i did an extensive search to find some tutorial on how to debug transformations and found nothing.

**Comment by 9037568** [comment]
I would suggest taking a look at how the Exec binding [does its transforms](https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/internal/ExecBinding.java#L148).

File: bundles/binding/org.openhab.binding.serial/README.md [hunk 4]
```diff
@@ -49,13 +49,14 @@ This binding does not have a configuration.
...
 
-The format has following variations:
+The format has the following variations:
 
 ```
 serial="<port>@<baudrate>" 
 serial="<port>@<baudrate>,REGEX(<regular expression>)" 
 serial="<port>@<baudrate>,BASE64 
 serial="<port>@<baudrate>,ON(<On string>),OFF(<Off string>)" 
+serial="<port>@<baudrate>,REGEX(<regular expression>), UP(<Up string>),Down(<Down string>), Stop(<Stop string>)"
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.serial/README.md]: For consistency, the down and stop should be in all caps:
DOWN(...)
STOP(...)

**Comment by marekhalmo** [comment]
Ok.. done, Is there anything else?

**Comment by marekhalmo** [comment]
I understand how the transformation works, i just don't see it installed in the Development environment and therefore i can't test the REGEX transformation used in the Serial binding. It is a functionality i did not change so there is no reason i must test it because of this pull request, but i wanted to test my Arduino serial gateway with the new support for Rollershutter. Could you somehow explain how to install the REGEX transformation in Development environment? I just can't see any transformations tab in paper UI

**Comment by marekhalmo** [comment]
Ok.. so i tested an example REGEX transformation in a rule..
rule "Exec rule check"
when
    Item ExampleSwitch changed
then
    transform("REGEX", "STATUS=(.*)", "STATUS=OK")
end

This executes and stops on a breakpoint in ExecTransformationService.transform(...) method.
So the issue is probably in the configuration of the transformation service in SerialBinding. I have seen some issue report on this subject. So i will write there.

Unless you have other change requests i consider this done. Let me know what's next.

Thank you!

**Comment by marekhalmo** [comment]
Awesome! 
When will this be available for download from Paper UI?
Thank you!

**Comment by 9037568** [comment]
We'll wait a few days to see if anyone has additional feedback before we merge it in.

**Comment by marekhalmo** [comment]
Well there has been a discussion about the regex transformation [here](https://github.com/openhab/openhab-core/issues/128). Should i make a fix into this branch?

**Comment by 9037568** [comment]
I think it'd be better to have a separate PR for that.

**Comment by marekhalmo** [comment]
Ok.. i will wait until​ this gets merged to avoid conflicts

**Comment by marekhalmo** [comment]
Hello, .. it's been a while, could you merge this so we can work on the regex issue?

**Comment by 9037568** [comment]
Thanks, @marekhalmo !

**Comment by marekhalmo** [comment]
Thank you! I will fix the RegexTransformationService issue for OH2 :)

**Comment by marekhalmo** [comment]
Hello, when will this be available for download in paperui?
Also i see no change yet on this page:
https://github.com/openhab/openhab1-addons/wiki/serial-binding

**Comment by marekhalmo** [comment]
This should probably change too
http://docs.openhab.org/addons/bindings/serial1/readme.html

**Comment by 9037568** [comment]
Did you [read up here?](http://docs.openhab.org/addons/bindings.html) I don't have much experience with OH2, but it looks like there are some configuration steps you have to do...

The first link - the [OH1 wiki](https://github.com/openhab/openhab1-addons/wiki/serial-binding) - only changes when someone changes it. You could update it, if you like.

The second link - the new docs - should come from the README.md that you already updated, but I'm not sure whether that happens automatically or when.

**Comment by marekhalmo** [comment]
Yes i did..

I think there has to be a new release of oh1 addons.. so i guess i have to wait until someone creates a new release. 
Until then i will just pus [the jar to addons](http://www.maniacs.sk/oh/org.openhab.binding.serial_1.10.0.201705031349.jar)

**Pull request merged**